### ISSUE TITLE
angular-cli: Extensibility

### ISSUE THREAD

**Issue by fkolar** [angular/angular-cli#2906]
This is more like  feature and maybe this is already in the list. 
Even all works great, I would like to extends angular-cli and have some hooks available, to I can plug in my custom-command that can be called everytime I spin a build, or anything while on ng serve I change something and the build refreshes. 

ideally the command would be outside of node_modules/

Thanks, 
-Fk

**Comment by grizzm0** [comment]
You can create scripts in package.json that chains multiple commands together. And live-reload on ng serve is already a thing...

**Comment by fkolar** [comment]
Right lets say I can use package.json from where I will execute a command, but does not help with live-reload - when running ng serve

**Comment by filipesilva** [comment]
We plan to support addons in the future, yes. It's a post 1.0 concern at the moment.

**Comment by tuvokki** [comment]
As I understand the cli is based on ember-cli, which supports the `install` action for installing addons:
```
Ember CLI will detect the presence of an addon by inspecting each of your applications dependencies and search their package.json files for the presence of ember-addon in the keywords section
```

Is it possible to enable this?

**Comment by filipesilva** [comment]
Just wanted to give an update: 1.0 is currently released, and we are looking at the design of an addon system for future major versions.

**Comment by fkolar** [comment]
This is good news!!  Is this going to be under wiki to look at ?

**Comment by fkolar** [comment]
Are there any news on this ?

**Comment by filipesilva** [comment]
We just added customizable blueprints (https://github.com/angular/angular-cli/pull/7090, docs are still being created) so that bit is customizable, but there isn't any build step customization yet.

**Comment by fkolar** [comment]
I see. I will just have to wait for this addons support that maybe allows me to extends or customize build steps. Any timeline for this?

But thanks for the prompt reply.

**Comment by filipesilva** [comment]
No time to share yet, no. But will try to share it as soon as we can.

**Comment by ShadowManu** [comment]
I would like/recommend the team to be more open about the addon system. The feature has been postponed and delayed for several reason (you know them better) but let the community help you out! Trying to get it right just by yourselves will not produce as good results. Therefore, some recommendations:

- An open process of whats the strategy in mind to allow extensibility, so people get an idea of what/how something can (or not) be done.

- Open a space for discussion of it. Probably the community can provide some insights (particularly webpack-oriented folks) on how to make it better/simpler. If the addons ecosystem is for the community to populate, then hear what the community thinks/need and let us help you out on the design (implementation?) process.

- Establish a public rough roadmap (with possible risks if needed). Some people exasperate (myself included in some stressful situations) when it hasn't been clear how much work/time is needed/used for the addon system. Since pre-1.0 release the standard response for extension issues has been "later". At least pinpointing which quarter of the year something is to be expected helps a lot.

I'm really enthusiastic to help on pug (jade) templating and service worker addons but I need you guys to be more open. Help me, help you, win-win :)

**Comment by Cito** [comment]
An addon system [has been promised](./1656#issuecomment-239366723) as solution for extending the default webpack configuration. Currently, I cannot even do such a simple thing as adding a loader for *graphql files without hacking node_modules/@angular/cli/models/webpack-configs/common.js.

If it will take too long to provide a proper addon system, it would be helpful to re-consider #1656 and at least provide an interim way of injecting webpack loaders using the angular-cli.json cofiguration. I really don't want to eject angular-cli just to be able to import *.graphql files.

**Comment by fkolar** [comment]
Yeah. I am hacking too few files and next to  it I have to have my own CLI script that initialize a few things ;-(

**Comment by chadbr** [comment]
@filipesilva 

Any updates / branches I could look into?

I'm basically at a standstill on some authorization workflows I need to do...

I'm not sure if I should fork / maintain merges or... what.

Thanks - Chad

**Comment by 4F2E4A2E** [comment]
Wow! This is happening!? Woot! Woot! Godspeed @filipesilva 👍

**Comment by alfaproject** [comment]
Sucks not having an easy way to add a loader for `.graphql` files. ):

**Comment by jonahbron** [comment]
Having pre-build hooks would be really nice for compiling GRPC-Web stubs.

**Comment by hansl** [comment]
I'm going to close this as Architect and Schematics are covering those bases. Architect is going to be stable in 7.0.

**Comment by chadbr** [comment]
@hansl great stuff by the way - congrats!
